Support for window.fetch

wmhilton opened the pull request:
I'm starting to work on issue #182 . ~~It's very basic right now. , and I haven't tested to see if it works. I also can't get `npm test` to run on my machine so I'm relying on Travis for feedback.~~

TODOs
- [x] Configuration option `preferXHR`: Use XHR instead of Fetch
- [ ] Test factory: Create fetch / non-fetch variants.
- [ ] rename to HttpRequestFS?
- [ ] update documentation

wmhilton commented:
Oh what fun: your Travis is running on precise and mine is running on trusty - your Firefox 30 didn't have "fetch" and the tests failed, my Firefox 50 does have fetch but failed for some other reason. This is really great for testing!

jvilk commented:
@wmhilton isn't it, though? 😄 That's why I have it all configured!

Why isn't `npm test` running? It should run on all major operating systems (Mac/Linux/Windows). That should be filed as a bug!

jvilk commented:
I added some things to your TODOs. For the purpose of testing, we should have an option to explicitly disable fetch that defaults to whether or not fetch is available (e.g. `!!global.fetch`).

That means the selection of which helper to use from `xhr.ts` will be decided by the file system, and not at initialization time like it is now. To avoid redundant conditionals (e.g. `if (shouldFetch)` all over the code), you can add `private` fields to the file system that save a reference to the chosen method of fetching / getting the size of files at file system construction time.

wmhilton commented:
Newly refactored code:
- Adds a `use` option to the `XmlHttpRequestOptions` interface with values `'auto'`, `'xhr'`, and `'fetch'`. ("use" feels a little *too* generic, but I couldn't think of a better one at the moment. `requestType`? `underlyingTech`?)
- Modifies `isAvailable()` so it will return true in environments with XmlHttpRequest OR fetch
- Adds a new private field `_useFetch` to remember selection
- Moved `fetchFileAsync` and `fetchFileSizeAsync` out of `xhr.ts` and into `generics/fetch.ts`

jvilk commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -97,6 +104,7 @@ export default class XmlHttpRequest extends BaseFileSystem implements FileSystem
 
   public readonly prefixUrl: string;
   private _index: FileIndex<{}>;
+  private _useFetch: boolean;
```
As in previous comments, it would be cleaner to replace this with function properties for downloading files and getting their size.

wmhilton replied:
> As in previous comments,

These must not be showing up in the order you wrote them. :/ Oh well, I'll figure it out eventually! It's like conversations on Twitter...

jvilk commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -396,7 +414,11 @@ export default class XmlHttpRequest extends BaseFileSystem implements FileSystem
   private _requestFileAsync(p: string, type: string, cb: BFSCallback<any>): void;
   private _requestFileAsync(p: string, type: string, cb: BFSCallback<any>): void {
-    asyncDownloadFile(this.getXhrPath(p), type, cb);
+    if (this._useFetch) {
```
It may be verbose to type out the full type of `_asyncDownloadFile`, so you can make an interface instead.

```typescript
interface DownloadFileMethod {
  (p: string, type: 'buffer', cb: BFSCallback<Buffer>): void;
  (p: string, type: 'json', cb: BFSCallback<any>): void;
  (p: string, type: string, cb: BFSCallback<any>): void;
}

// etc

private _asyncDownloadFile: DownloadFileMethod;
```

wmhilton replied:
> I mentioned this elsewhere, but you can avoid all of these conditionals

all two? Am I missing some somewhere?

jvilk replied:
Oh, I didn't mean for that to sound sassy here. I'm referring to [this comment](https://github.com/jvilk/BrowserFS/pull/185#issuecomment-319266217):

> To avoid redundant conditionals (e.g. if (shouldFetch) all over the code), you can add private fields to the file system that save a reference to the chosen method of fetching / getting the size of files at file system construction time.

wmhilton replied:
No worries! I didn't think you were being sassy. I was genuinely wondering if there were more than two places that needed that conditional.

jvilk commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -69,11 +75,12 @@ export default class XmlHttpRequest extends BaseFileSystem implements FileSystem
       XmlHttpRequest.FromURL(opts.index, cb, opts.baseUrl, false);
     } else {
-      cb(null, new XmlHttpRequest(opts.index, opts.baseUrl, false));
+      cb(null, new XmlHttpRequest(opts.index, opts.baseUrl, false, opts.use));
```
You'll need to surround this with a try / catch since the constructor throws an exception now.

```typescript
let xhr: XmlHttpRequest;
try {
  xhr = new XmlHttpRequest(opts.index, opts.baseUrl, false, opts.use);
} catch (e) {
  return cb(e);
}
cb(null, xhr);
```

Note that surrounding the `cb` in the `try` / `catch` might accidentally catch & swallow errors thrown by user code. (I'm sure I'm forgetting to follow this advice elsewhere in the codebase, but then that's my error!)

wmhilton replied:
constructor no longer throws an exception (since I am using a boolean instead of a string), so... wait, no, it still can throw an exception if `listingUrlOrObj` is a string! I guess that was a pre-existing "bug". I think this means we'll need to catch and handle construction errors in both `.Create` and in `.FromURL`.

wmhilton replied:
So technically, I think adding try/catch to .FromURL would *also* prevent any unhandled errors in .Create... but I'm going to add try/catch to .Create as well just so we don't shoot ourselves in the foot later if we add another kind of thrown error to the constructor.

jvilk replied:
The constructor is private in V2.0 and users should not call it directly. `listingUrlOrObj` as a string in the constructor is deprecated and is removed in V2.0. `Create` handles that variant like it does in V1.x -- asynchronously.

So, don't worry about that exception. This is one of the motivations for the `Create` method -- keeping constructors simple, and centralizing the checking of user-passed options!

jvilk replied:
[For context, this is XHR in the V2 branch](https://github.com/jvilk/BrowserFS/blob/dropbox_v2/src/backend/XmlHttpRequest.ts#L88). 

(I won't make you rebase; I can handle the merge into V2 myself.)

jvilk commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -69,11 +75,12 @@ export default class XmlHttpRequest extends BaseFileSystem implements FileSystem
   public static isAvailable(): boolean {
-    return typeof(XMLHttpRequest) !== "undefined" && XMLHttpRequest !== null;
+    return (typeof(fetch) !== "undefined" && fetch !== null) ||
+           (typeof(XMLHttpRequest) !== "undefined" && XMLHttpRequest !== null);
```
Note: `supportsSynch()` needs to be updated in the Node case. If `XMLHttpRequest` isn't available, there is no way to synchronously download files, and `supportsSynch()` should return `false`.

wmhilton replied:
Ooh. That does lead to an interesting scenario where fetch is used for async calls but then XHR is used for sync calls. Slightly more complicated than I thought.

jvilk replied:
Yeah. This is why I'm now wondering if the abstract class route might be cleaner.

jvilk commented on src/generic/fetch.ts:
```diff
@@ -0,0 +1,57 @@
+              .catch((err) => cb(new ApiError(0, err.message)));
+            break;
+          default:
+            throw new ApiError(ErrorCode.EINVAL, "Invalid download type: " + type);
```
You should call the callback with this error instead of throwing it. (Do I `throw` the error in `xhr.ts`? That should be fixed if I do.)

wmhilton replied:
Looks like `xhr.ts` is fine! I think I just threw out of habit.

jvilk commented on src/generic/fetch.ts:
```diff
@@ -0,0 +1,57 @@
+  fetch(p)
+    .then((res) => {
+      if (!res.ok) {
+        return cb(new ApiError(res.status, "fetch error."));
```
Can you change this to use `ErrorCode.EIO`? I realize that I did this in my `xhr` code, but that should really be fixed. The error code should always be one of those libc error codes. You can include the status code in the error message, e.g.:

```typescript
new ApiError(ErrorCode.EIO, `fetch error: response returned code ${res.status}`)
```

If you could update my XHR code to do the same, I'd be very grateful. I haven't looked over that part of the code in years!

wmhilton replied:
Haha! I *wondered* about that. I thought, "That's so weird... he's not using his fancy Error type here, must be some reason... I better copy his behavior though!" EIO sounds like a good generic error choice.

wmhilton replied:
Will changing the error code from status to EIO break any existing applications? I'm wondering if it's a breaking change.

jvilk replied:
It shouldn't! This should only happen:

* If the file index is incorrect, and refers to a file that doesn't actually exist.
* If the network goes out, preventing network requests from working.

The unit tests do not test either of these occurrences. The existing error codes are nonsensical to any code written against the API. Most code running with BrowserFS is written in a different programming language/for the native environment (C/C++/Java/Node JS/etc) and expects the standard libc error codes.

In both cases, `EIO` seems like the correct error code to me.

jvilk commented on src/generic/fetch.ts:
```diff
@@ -0,0 +1,57 @@
+  fetch(p, { method: 'HEAD' })
+    .then((res) => {
+      if (!res.ok) {
+        return cb(new ApiError(res.status, "fetch HEAD error."));
```
Ditto with above!

jvilk commented on src/generic/fetch.ts:
```diff
@@ -0,0 +1,57 @@
+        }
+      }
+    })
+    .catch((err) => cb(new ApiError(0, err.message)));
```
Does the fetch API ever throw an error that needs to be `catch`'ed? I couldn't find the applicable part of the standard through a quick grep.

I'd use `ErrorCode.EIO` for the error code here.

wmhilton replied:
No, fetch is always asynchronous and always returns a Promise, so it never throws only rejects.

wmhilton replied:
As usual, I'm slightly wrong! There is one edge case: https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Exceptions

> TypeError: Since Firefox 43, fetch() will throw a TypeError if the URL has credentials, such as http://user:password@example.com.

Obviously that's an error that we'd want to bubble up to the user somehow. Should we wrap the whole thing in a try/catch and pass the error to the callback?

jvilk replied:
Oh, sorry for the ambiguity. I meant `.catch`'ed -- meaning, a rejected promise. (We shouldn't `try / catch` this; that seems like a truly exceptional error.)

I could not find anything in the spec (in my brief search) concerning when the promise gets rejected, and if it gets rejected with an error object.

Regardless, it seems like it should always be an `EIO` error anyway, as all of the other relevant checks occur on the file index.

wmhilton replied:
Ah! I see. Oops, I already wrapped it. `fetch` rejects for network errors: CORS errors, DNS lookup failures, TCP connection dropped, that sort of thing.

jvilk commented on src/generic/fetch.ts:
```diff
@@ -0,0 +1,57 @@
+        return cb(null, parseInt(res.headers.get('Content-Length') || '-1', 10));
+      }
+    })
+    .catch((err) => cb(new ApiError(0, err.message)));
```
Ditto with above!

wmhilton commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -31,6 +33,10 @@ export interface XmlHttpRequestOptions {
   baseUrl?: string;
+  // Whether to use Fetch or XmlHttpRequest.
+  // Can be 'xhr', 'fetch', or 'auto'.
+  // Default: 'auto' which uses fetch if its available, and XHR as a fallback
```
Eh... I kind of foresee that scenario. What if you want to use XHR, unless it's not available? I mean, that's basically where we are with Service Workers right? Both  "forceXHR" and a "forceFetch" could end up screwing us over if that feature is unavailable but the other one is. What about "preferXHR"? (where preferXHR = true means prefer XHR, fallback to fetch and preferXHR = false means prefer fetch, fallback to XHR)

But one reason I didn't make it a boolean was I foresee something after fetch. Maybe something with streams. I feel like using a boolean to define how to handle what is really 4 different cases (neither available, just xhr, just fetch, or both) is putting us at the limit of what a single boolean config can do and any future stuff will require adding a *new* option instead of more string variations.

wmhilton replied:
I'm making it a boolean. Fewer ways to screw up. :)

jvilk replied:
Oh, sorry for the delay! I see what you mean now. In the Node environment, it's also plausible that someone would want to use the `HTTP` API.

Another road to go, and maybe this is crazy, is to make separate `XmlHttpRequest`, `FetchFS`, etc file systems. Abstract away the core `HttpRequestFS` into an abstract class, and extend it with the `downloadAsync` / `downloadSync` / `getSizeAsync` / etc methods. This would also easily preserve backwards compatibility, and make it possible to exclude environment-specific file systems from e.g. the browser bundle.

jvilk replied:
You can stick with the boolean for now. I can do the fancy class maneuver if it makes sense post-merge.

wmhilton replied:
Yeah I hadn't even thought of it, but in Node someone might want to use `http`. I don't see *why* -
 I would just use `isomorphic-fetch` in that scenario - but I'm sure there are cases. Maybe to enable streaming? Or proxying. Or performance reasons or something.

Yes, we have an abstraction conundrum. One solution with two cases, or two solutions? If we separate XmlHttpRequest and FetchFS, then the user will be responsible for handling it, and have to learn and adapt to the oddities like: there's no XmlHttpRequest in ServiceWorker *but* Fetch doesn't have a synchronous option. I think having both together makes it easiest on the user. There are not so many possibilities that we can't provide an optimal experience for every use case. I like the current behavior, which lets you use sync or async file system operations, but uses fetch when it can. So in Service Workers sync operations will not be supported, but that's the only catch.

wmhilton replied:
> but uses fetch when it can

^ unless you've specified `preferXHR`

jvilk replied:
(I actually would like to support fetch's streaming downloads at some point to avoid prefetching entire files before any operations can proceed, but that's a later PR.)

I'll think on this some more.

wmhilton commented:
New changes:
- renames `use` to `preferXHR` and makes it a boolean
- modifies `supportsSynch()` to return false if XMLHttpRequest is unavailable
- moves the definitions for `_requestFileAsync` and `_requestFileAsync` to the constructor function
- fixes the error handling in `fetch.ts`

jvilk commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -136,6 +153,14 @@ export default class XmlHttpRequest extends BaseFileSystem implements FileSystem
     this._index = FileIndex.fromListing(listing);
+
+    this._requestFileAsync = (preferXHR || !('fetch' in global))
+      ? (p: string, type: string, cb: BFSCallback<any>) => asyncDownloadFile(this.getXhrPath(p), type, cb)
```
`(p, type, cb) => asyncDownloadFile(this.getXhrPath(p), type, cb)` gets translated into the following JavaScript code:

```javascript
var _this = this;
function(p, type, cb) { asyncDownloadFile(_this.getXhrPath(p), type, cb); }
```

Since `_requestFileAsync` is always called on the `this` object, you can use an ordinary `function` and tell TypeScript the expected type of `this`:

```typescript
// Note: 'this' is a meta-argument
function(this: XmlHttpRequest, p: string, type: string, cb: BFSCallback<any>) { return asyncDownloadFile(this.getXhrPath(p), type, cb); }
```

JavaScript:

```javascript
function(p, type, cb) { return asyncDownloadFile(this.getXhrPath(p), type, cb); }
```

This code avoids an unneeded variable capture in the closure, and is a bit cleaner in the compiled code.

You don't have to do this for the PR to be accepted; I'll do it post-merge if it's annoying.

jvilk commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -167,7 +192,7 @@ export default class XmlHttpRequest extends BaseFileSystem implements FileSystem
 
   public supportsSynch(): boolean {
-    return true;
+    return (typeof(XMLHttpRequest) !== "undefined" && XMLHttpRequest !== null);
```
What do you think about adding this and the `fetch` check into `xhr.ts` and `fetch.ts` as exported `const` booleans?

e.g.

```typescript
export const xhrIsAvailable = typeof(XMLHttpRequest) !== "undefined" && XMLHttpRequest !== null;
```

(Like the other comment, I can do this post-merge.)

wmhilton replied:
I think that's an excellent idea.

jvilk commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -390,16 +415,6 @@ export default class XmlHttpRequest extends BaseFileSystem implements FileSystem
-  /**
    * Synchronously download the given file.
    */
   private _requestFileSync(p: string, type: 'buffer'): Buffer;
```
Another thing you could choose to tackle, or I can do post-merge: make these into properties, and replace with stubs that throw an [`ENOTSUP`](https://github.com/jvilk/BrowserFS/blob/master/src/core/file_system.ts#L757) ApiError when XHR isn't available.

As it stands, calling a synchronous method on this file system in e.g. Node will try to call XHR and will fail with a more mysterious error.

wmhilton replied:
Mmm. You're right, I think that's a good idea. I thought that sync filesystem functions would be using the `supportsSynch` property to check whether the fs supported sync functions before calling them? If not then yes we should definitely stub these for the no-XHR case.

jvilk replied:
I looked through the code, and it doesn't. I added methods like `supportsSynch` primarily for unit tests, and secondarily for users of the library that need to generically check what operations a backend supports. I relied on the `BaseFileSystem` throwing `ENOTSUP` errors for sync operations on async file systems, and never anticipated that a sync backend might be async in some cases. Plus, there are some cases where synchronous operations may or may not be supported depending on the path (e.g., a MountableFileSystem with sync and async backends).

jvilk commented:
(I added two more super picky comments that you don't have to address.)

wmhilton commented:
I'm sure this is just my lack of familiarity with karma, but I can't figure out what/where/how to modify the tests to add a `"preferXHR: false"` variant. Any chance you could add that? Also, what do you think of the current code? I think I addressed all your feedback points, but you might still find ways to improve it.

jvilk commented:
It's not in Karma! There are TypeScript files that create multiple variants of each file system for testing (hence, they are 'factories').

The following is the one for XHRFS, which currently has two variants (listing URL passed to `Create()` which is async, listing URL passed to `constructor` which is sync but will be removed in next major revision):

https://github.com/jvilk/BrowserFS/blob/master/test/harness/factories/xhrfs_factory.ts#L4

I'd like you to modify the test so that:

* The existing two XHR file systems prefer XHR, preserving the existing behavior.
* Add a third file system configuration that does not prefer XHR.

Make sense?

jvilk commented:
@wmhilton I just did the big 2.0 merge. I promise that's the last churn on this pull request!

This feature will go into the first 2.0 release.

Let me know if you've given up on the final few changes, and I will solve the merge conflicts and add the needed test changes.

wmhilton commented:
> Let me know if you've given up on the final few changes, and I will solve the merge conflicts and add the needed test changes.

Could you? I keep squinting at the code in `xhrfs_factory.ts` and I see how there are branches (the if statements) but I don't actually see how that turns into two test variants. Wait, unless... is the `cb` being used to append a test to the end of a hidden queue rather than being a continuation? 🤔 

Um... I'm sorry @jvilk I just don't have enough experience for it to click. If you add the test, and then I look at the diff, I'm sure I'll slap my head and say "doh!" but right now I'd just be shooting in the dark.

jvilk commented:
@wmhilton All merged! You can check out the changes I made if you want. I'll get on the other tasks before the 2.0 release (e.g. the rename, maintaining backwards compat, etc).
